pdns/syncres.cc

```diff
@@ -4459,3 +4461,161 @@ int SyncRes::getRootNS(struct timeval now, asyncresolve_t asyncCallback, unsigne
+
+#ifdef HAVE_LIBCURL
+
+  MiniCurl mc;
+  string reply = mc.getURL(url, nullptr, nullptr, 10, false, true);
+  vector<string> lines;
+  std::istringstream stream(reply);
+  string line;
+  while (std::getline(stream, line)) {
+    lines.push_back(line);
+  }
+
+  const DNSName zone(name);
+  DNSResourceRecord drr;
+  DNSName qname;
+  QType qtype;
+  vector<DNSRecord> v;
+  vector<DNSRecord> sigs;
+  vector<shared_ptr<RRSIGRecordContent>> sigsrr;
+  bool first = true;
+  time_t now = time(nullptr);
+  recordsAndSignatures soa;
+  uint32_t last_ttl = 0;
+
+  const string msg  = "zones-to-cache error while loading " + name + " from " + url + ": ";
+  ZoneParserTNG zpt(lines, zone);
+  zpt.setMaxGenerateSteps(0);
+
+  // The code below assumes the records for a name,type tuple are grouped together and followed by their corresponding sigs.
+  // We do not do validation, it will happen on-demand if an Indeterminate record is encountered when the caches are queried
+  while (zpt.get(drr)) {
+    DNSRecord dr(drr);
+    if (first) {
+      qname = dr.d_name;
+      qtype = dr.d_type;
+      first = false;
+    }
+    if (qname != dr.d_name || qtype != dr.d_type) {
+      // We came to a different block
+      if (qname == dr.d_name && dr.d_type == QType::RRSIG) {
+        // same name,, but now collect sigs for it
+        qtype = QType::RRSIG;
+      } else {
+        // done with the previous block of records and corresponding sigs, store them into the caches
+        if (qtype == QType::SOA) {
+          if (soa.records.size() > 0) {
+            g_log << Logger::Error << msg << "can only load zones with a single SOA record ";
+            return refresh;
+          }
+          soa.records = v;
+          if (dnssec) {
+            soa.signatures = sigs;
+          }
+        }
+        switch (qtype) {
+        case QType::NSEC:
+        case QType::NSEC3: {
+          if (soa.records.size() == 0) {
+            g_log << Logger::Error << msg << "SOA record should precede any NSEC/NSEC3 records";
+            return refresh;
+          }
+          NegCache::NegCacheEntry ne;
+          ne.authoritySOA = soa;
+          ne.DNSSECRecords.records = v;
+          ne.DNSSECRecords.signatures = sigs;
+          ne.d_name = qname;
+          ne.d_auth = zone;
+          ne.d_ttd = now + last_ttl;
+          ne.d_qtype = qtype;
+          g_negCache->add(ne);
+
+#if 0 // Do not do this yet, I'm not sure it is correct and it will get fiLled on-demand anyway
+          if (g_aggressiveNSECCache && v.size() > 0) {
+            vState state = validateRecordsWithSigs(0, qname, qtype, qname, qtype, v, sigsrr);
+            if (state == vState::Secure) {
+              g_aggressiveNSECCache->insertNSEC(zone, qname, v.at(0), sigsrr, qtype == QType::NSEC3);
+            }
+          }
+#endif
+          break;
+        }
+        default:
+          if (!dnssec) {
+            sigsrr.clear();
+          }
+          bool auth = qname == zone;
```

Comment:
That doesn't sound right, but we would probably need to keep track of delegations to know which records are authoritative?

Reply:
This indeed a quick hack that works for (most of?) the records as they appear in the root zone. I should have commented on that.

Reply:
Right, but in that case perhaps we need to document that it only works with a root zone at the moment?

Reply:
And if we are indeed supporting the root zone only, it might make sense to treat these records as Secure and to insert the NSEC into the NSEC cache?